File: p2p-java/src/Account.java
//import com.sun.xml.internal.ws.api.message.ExceptionHasMessage;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.InetAddress;
import java.net.URL;
import java.net.UnknownHostException;
import java.util.Scanner;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.ReentrantLock;

public class Account extends Node{
    private Double amount;
    private String name;
    private ReentrantLock mutex;
    private String filename;

    public Account(String host, int port, String name) throws Exception {
        super(host, port);
        //We will set the starting amount as 0
        this.amount = 0.0;
        this.name = name;
        this.filename = name+".txt";
        File file = new File(this.filename);
        Boolean iscreated = file.createNewFile();
        if(iscreated){
            FileWriter myWriter = new FileWriter(this.filename);
            myWriter.write("0.0");
            myWriter.close();
        }else {
            //We extract the amount from the txt file and set it to the balance
            Scanner myReader = new Scanner(file);
            this.amount = Double.valueOf(myReader.nextLine());
        }
        this.mutex = new ReentrantLock();
        String response = this.post_request(name, host, Integer.toString(port));
        System.out.println(response);
    }

    public int moneyTransfer(String username, Double amount) throws Exception {
        String res = get_request(username);
        if(res == null){
            System.out.println("Error! Receiving user not found!");
            return -1;
        }
        String __host = res.split(":")[0];
        String __port = res.split(":")[1];
        return this.send_money(__host, Integer.parseInt(__port), amount);
    }

    public String get_request(String username) throws Exception {
        URL url = new URL("https://evening-headland-54924.herokuapp.com/username?username=" + username);
        HttpURLConnection con = (HttpURLConnection) url.openConnection();
        con.setRequestMethod("GET");

        int status = con.getResponseCode();
        if(status>299){
            System.out.println("Changing server!...");
            URL urlSec = new URL("https://fast-atoll-78029.herokuapp.com/username?username=" + username);
            HttpURLConnection conSec = (HttpURLConnection) urlSec.openConnection();
            conSec.setRequestMethod("GET");

            BufferedReader inSec = new BufferedReader(new InputStreamReader(con.getInputStream()));
            String inputLineSec;
            StringBuffer contentSec = new StringBuffer();
            while ((inputLineSec = inSec.readLine()) != null) {
                contentSec.append(inputLineSec);
            }
            inSec.close();
            if(contentSec.toString() == null){
                throw new Exception("User couldn't be found out");
            }
            return contentSec.toString();
        }

        BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()));
        String inputLine;
        StringBuffer content = new StringBuffer();
        while ((inputLine = in.readLine()) != null) {
            content.append(inputLine);
        }
        in.close();
        if(content.toString() == null){
            throw new Exception("User couldn't be found out");
        }
        return content.toString();
    }

    public String post_request(String username, String ip, String port) throws Exception {
        URL url = new URL("https://evening-headland-54924.herokuapp.com/add?username=" + username + "&ip=" +ip+ "&port="+port);
        System.out.println(url);
        HttpURLConnection con = (HttpURLConnection) url.openConnection();
        con.setRequestMethod("POST");

        int status = con.getResponseCode();
        if(status > 299){
            BufferedReader inSec = new BufferedReader(new InputStreamReader(con.getInputStream()));
            String inputLineSec;
            StringBuffer contentSec = new StringBuffer();
            while ((inputLineSec = inSec.readLine()) != null) {
                contentSec.append(inputLineSec);
            }
            inSec.close();
            if(contentSec.toString().equals("Err")){
                //User already exists
                throw new Exception("User already exists");
            }
            return contentSec.toString();
        }

        BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()));
        String inputLine;
        StringBuffer content = new StringBuffer();
        while ((inputLine = in.readLine()) != null) {
            content.append(inputLine);
        }
        in.close();
        if(content.toString().equals("Err")){
            //User already exists
            throw new Exception("User already exists");
        }
        return content.toString();
    }

    public int send_money(String account_ip, int account_port, Double amount){
        String message;
        //First we check that we have enough money to send this amount
        if(amount < 0.0){
            System.out.println("Error! You can't deposit a negative sum into an account.");
            return -1;
        }
        if(this.amount < amount){
            System.out.println("Money transfer to Account: " + account_ip + ":" + account_port + "failed!: Insufficient Funds!");
            return -1;
        }
        message = Double.toString(amount);
        try{
            this.mutex.lock();
            this.amount = this.amount - amount;
            FileWriter myWriter = new FileWriter(this.filename);
            myWriter.write(this.amount.toString());
            myWriter.close();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            this.mutex.unlock();
        }
        this.send_message(message, account_ip, account_port);
        return 0;
    }

    @Override
    public void addAmount(String amount){
        //Convert the amount to a Double value
        Double amount_d = Double.parseDouble(amount);
        if(amount_d < 0){
            System.out.println("Error! You can't add a negative sum of money into an account.");
            return;
        }
        System.out.println("Adding " + amount_d.toString() + " € to your account!");
        //Here we will implement the mutex lock
        try{
            this.mutex.lock();
            //Here the update of the balance occurs
            this.amount += amount_d;
            FileWriter myWriter = new FileWriter(this.filename);
            myWriter.write(this.amount.toString());
            myWriter.close();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            this.mutex.unlock();
        }
    }

    public void removeAmount(Double amount){
        if(amount < 0){
            System.out.println("Error! You can't remove a negative sum of money into an account.");
            return;
        }
        System.out.println("Taking out " + amount.toString() + " € from your account!");
        //Here we will implement the mutex lock
        try{
            this.mutex.lock();
            this.amount = this.amount - amount;
            FileWriter myWriter = new FileWriter(this.filename);
            myWriter.write(this.amount.toString());
            myWriter.close();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            this.mutex.unlock();
        }
    }

    public void depositAmount(Double amount){
        if(amount < 0){
            System.out.println("Error! You can't deposit a negative sum into an account.");
            return;
        }
        System.out.println("Initiating sample deposit of money into account...");
        this.addAmount(amount.toString());
    }

    public static boolean confirmation(String s) throws UnknownHostException {
        String prove = "";
        String aux = InetAddress.getLocalHost() + "";
        System.out.println(aux);
        prove += aux.substring(aux.indexOf("/")+1);
        if (s.equals(prove)){
            System.out.println("Valid IP direction");
            return true;
        }
        System.out.println("Invalid IP direction");
        return false;
    }

    public static void main(String[] args) {
        Account accountDest = null;
        try {
            accountDest = new Account("127.0.0.1", 19011, "destinationUser");
        } catch (Exception e) {
            e.printStackTrace();
        }


        Account account1 = null;
        try {
            account1 = new Account("127.0.0.1", 18082, "usuarioprueba1");
        } catch (Exception e) {
            e.printStackTrace();
        }

        Account account2 = null;
        try {
            account2 = new Account("127.0.0.1", 18083, "usuarioprueba2");
        } catch (Exception e) {
            e.printStackTrace();
        }

        Thread t1 = new Thread(accountDest);
        t1.start();

        final Account finalAccount1 = account1;
        Thread tsend1 = new Thread(new Runnable() {
            @Override
            public void run() {
                assert finalAccount1 != null;
                try {
                    finalAccount1.moneyTransfer("destinationUser", 10.0);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });

        final Account finalAccount2 = account2;
        Thread tsend2 = new Thread(new Runnable() {
            @Override
            public void run() {
                assert finalAccount2 != null;
                try {
                    finalAccount2.moneyTransfer("destinationUser", 10.0);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });

        tsend1.start();
        tsend2.start();

        try {
            TimeUnit.SECONDS.sleep(5);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        t1.interrupt();
        accountDest.stop_node();


    }

    private Double getAmount() {
        return this.amount;
    }
}

File: p2p-python/account.py
from node import Node
import requests
import os.path

class Account(Node):

    def __init__(self, host, port, name):
        super(Account, self).__init__(host, port)
        self.filename = name+".txt"
        if os.path.isfile(self.filename):
            #If the file already exists then we read the file to find the amount
            f = open(self.filename, "r")
            self.balance = float(f.read())
        else:
            #If the file doesn't exist we create it and write the initial balance of 0.0
            self.balance = 0.0
            f = open(self.filename, "w")
            f.write("0.0")
        print("Hola")
        f.close()
        self.name = name
        self.post_request(name, host, port)

    def post_request(self, username, ip, port):
        url = "https://evening-headland-54924.herokuapp.com/dsfsdfdf?username="+username+"&ip="+ip+"&port="+str(port)
        print(url)
        r = requests.post(url = url)
        print(r.status_code)
        if r.status_code > 299:
            print("Changing to secondary server...")
            urlAuxSecondary = "https://fast-atoll-78029.herokuapp.com/add?username="+username+"&ip="+ip+"&port="+str(port)
            resp = requests.post(url = urlAuxSecondary)
        if r == "Err":
            raise Exception("The user already exists in the system")

    def get_request(self,username):
        url = "https://evening-headland-54924.herokuapp.com/badssdasndsad?username="+username
        r = requests.get(url = url)
        if r is None:
            raise Exception("The user you are trying to send money to doesn't exist")
        if r.status_code > 299:
            print("Changing to secondary server in GET...")
            urlAuxSecondary = "https://fast-atoll-78029.herokuapp.com/username?username="+username
            resp = requests.get(url = urlAuxSecondary)
            data = resp.text
            ip = data.split(':')[0]
            port = data.split(':')[1]
            return ip, port

        data = r.text
        ip = data.split(':')[0]
        port = data.split(':')[1]
        return ip, port

    def transferMoney(self, username, amount):
        ip, port = self.get_request(username)
        if float(amount) > self.balance:
            raise Exception("You do not have sufficient funds to make this transfer")
        elif float(amount) <= 0:
            raise Exception("You can not transfer a negative or null amount of money")
        else:
            self.send_to_node(amount, ip, int(port))
            self.lock.acquire()
            self.balance = self.balance - float(amount)
            f = open(self.filename, "w")
            f.write(self.balance)
            f.close()
            self.lock.release()

    def node_message(self, data):
        '''
        node_message: This function is called when a node receives a text message,
            in it we will raise a lock, and update the balance of the user's Account
            data: String of the data received by the node.
        '''
        amount = float(data)
        self.lock.acquire()
        print("Lock acquired!")
        self.balance += amount
        f = open(self.filename, "w")
        f.write(self.balance)
        f.close()
        self.lock.release()
        print("Lock released!")
        print("Your resulting balance after the operation is: " + str(self.balance))
        
    